### PR TITLE
[MRG] CO-Optimal Transport solver

### PR DESCRIPTION
**Pull request by 6Ulm**
## Types of changes

Implementation of CO-Optimal Transport (COOT) [47] in the new file `coot.py`, containing two methods: 
- `co_optimal_transport`: returns the sample and feature couplings.
- `co_optimal_transport2`: returns the COOT distance.

Both methods can handle:
- Unregularized COOT ($\varepsilon = 0$). In this case, `ot.lp.emd` solver will be used.
- Entropic regularized COOT ($\varepsilon > 0$). In this case, `ot.sinkhorn` solver will be used.
- COOT with linear terms for both sample and feature couplings, in the same spirit as fused Gromov-Wasserstein (in which only linear term for sample coupling exists)

The COOT distance outputed by `co_optimal_transport2` is also sub-differentiable with respect to the input matrices and marginal distributions.

## Motivation and context / Related issue

CO-Optimal Transport is not yet available on POT.

## How has this been tested (if it applies)

- The implementation is tested sucessfully on toy examples. The tests are reported in `test_coot.py`.
- Tests on more complex examples are reported in examples.

## PR checklist

- [x] I have read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR and Issue fix to the [**RELEASES.md**](RELEASES.md) file.

### CODE REVIEW

**Review by agramfort — APPROVED** [review]
pushed a few fixes to the Doc

**Comment by agramfort** [comment]
no idea what's happening with the doc here...

File: README.md [hunk 1]
```diff
@@ -303,6 +303,8 @@ Dictionary Learning](https://arxiv.org/pdf/2102.06555.pdf), International Confer
 
 [46] Bonet, C., Berg, P., Courty, N., Septier, F., Drumetz, L., & Pham, M. T. (2023). [Spherical Sliced-Wasserstein](https://openreview.net/forum?id=jXQ0ipgMdU). International Conference on Learning Representations.
 
-[47] Chowdhury, S., & Mémoli, F. (2019). [The gromov–wasserstein distance between networks and stable network invariants](https://academic.oup.com/imaiai/article/8/4/757/5627736). Information and Inference: A Journal of the IMA, 8(4), 757-787.
+[47] Redko, I., Vayer, T., Flamary, R., and Courty, N. (2020). [CO-Optimal Transport](https://proceedings.neurips.cc/paper/2020/file/cc384c68ad503482fb24e6d1e3b512ae-Paper.pdf). Advances in Neural Information Processing Systems, 33.
```
Comment (rflamary) [inline README.md:306]: add your ref at the end and use the same number in the doc of the function

File: ot/coot.py [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+    """
+
+    def compute_kl(p, q):
+        kl = nx.sum(p * nx.log(p + 1.0 * (p == 0))) - nx.sum(p * nx.log(q))
+        return kl
+
+    # Main function
+
+    if method_sinkhorn not in ["sinkhorn", "sinkhorn_log"]:
+        raise ValueError(
```
Comment (rflamary) [inline ot/coot.py:149]: make  tets that check that the error is raised

File: ot/coot.py [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+            "Method {} is not supported in CO-Optimal Transport.".format(method_sinkhorn))
+
+    X, Y = list_to_array(X, Y)
+    nx = get_backend(X, Y)
+
+    if isinstance(epsilon, float) or isinstance(epsilon, int):
+        eps_samp, eps_feat = epsilon, epsilon
+    else:
+        if len(epsilon) != 2:
+            raise ValueError("Epsilon must be either a scalar or an indexable object of length 2.")
```
Comment (rflamary) [inline ot/coot.py:159]: again test

File: ot/coot.py [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+        else:
+            eps_samp, eps_feat = epsilon[0], epsilon[1]
+
+    if isinstance(alpha, float) or isinstance(alpha, int):
+        alpha_samp, alpha_feat = alpha, alpha
+    else:
+        if len(alpha) != 2:
+            raise ValueError("Alpha must be either a scalar or an indexable object of length 2.")
```
Comment (rflamary) [inline ot/coot.py:167]: same here

File: ot/coot.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+    # initialize coupling and dual vectors
+    if warmstart is None:
+        pi_samp, pi_feat = wxy_samp, wxy_feat  # shape nx_samp x ny_samp and nx_feat x ny_feat
+        duals_samp = (nx.zeros(nx_samp, type_as=X), nx.zeros(
+            ny_samp, type_as=Y))  # shape nx_samp, ny_samp
+        duals_feat = (nx.zeros(nx_feat, type_as=X), nx.zeros(
+            ny_feat, type_as=Y))  # shape nx_feat, ny_feat
+    else:
+        pi_samp, pi_feat = warmstart["pi_sample"], warmstart["pi_feature"]
+        duals_samp, duals_feat = warmstart["duals_sample"], warmstart["duals_feature"]
```
Comment (rflamary) [inline ot/coot.py:208]: also add a warmstart in one of teh tests

File: ot/coot.py [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+                coot = coot + eps_samp * compute_kl(pi_samp, wxy_samp)
+            if eps_feat != 0:
+                coot = coot + eps_feat * compute_kl(pi_feat, wxy_feat)
+            list_coot.append(coot)
+
+            if err < tol_bcd or abs(list_coot[-2] - list_coot[-1]) < early_stopping_tol:
+                break
+
+            if verbose:
+                print(
```
Comment (rflamary) [inline ot/coot.py:257]: tets also with verbose

File: ot/coot.py [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+        2 * pi_samp @ Y @ pi_feat.T  # shape (nx_samp, nx_feat)
+    gradY = 2 * Y * (wy_samp[:, None] * wy_feat[None, :]) - \
+        2 * pi_samp.T @ X @ pi_feat  # shape (ny_samp, ny_feat)
+
+    coot = dict_log["distances"][-1]
+    coot = nx.set_gradients(coot, (wx_samp, wx_feat, wy_samp, wy_feat, X, Y),
+                            (vx_samp, vx_feat, vy_samp, vy_feat, gradX, gradY))
+
+    if log:
+        return coot, dict_log
```
Comment (rflamary) [inline ot/coot.py:431]: test both

**Review by rflamary — COMMENTED** [review]
Hello @6Ulm we are nearly there, could you pelae upadte teh tests as discussed above?

**Pull request merged**